libraries/AP_HAL_Linux/RCInput_RPI.cpp

```diff
@@ -355,23 +366,63 @@ void RCInput_RPI::init_ctrl_data()
     ((dma_cb_t *)con_blocks->get_page(con_blocks->_virt_pages, cbp))->next = (uintptr_t)con_blocks->get_page(con_blocks->_phys_pages, 0);
 }
 
+void RCInput_RPI::init_PCM()
+{
+    if (_version != 4)
+    {
+        init_PCM_BCM2835();
+    }
+    else
+    {
+        init_PCM_BCM2711();
+    }
+}
+
 /*Initialise PCM
   See BCM2835 documentation:
   http://www.raspberrypi.org/wp-content/uploads/2012/02/BCM2835-ARM-Peripherals.pdf
  */
-void RCInput_RPI::init_PCM()
+void RCInput_RPI::init_PCM_BCM2835()
 {
     pcm_reg[RCIN_RPI_PCM_CS_A] = 1;                                          // Disable Rx+Tx, Enable PCM block
     hal.scheduler->delay_microseconds(100);
     clk_reg[RCIN_RPI_PCMCLK_CNTL] = 0x5A000006;                              // Source=PLLD (500MHz)
     hal.scheduler->delay_microseconds(100);
-    clk_reg[RCIN_RPI_PCMCLK_DIV] = 0x5A000000 | ((50000/RCIN_RPI_SAMPLE_FREQ)<<12);   // Set pcm div. If we need to configure DMA frequency.
+    clk_reg[RCIN_RPI_PCMCLK_DIV] = 0x5A000000 | ((RCIN_RPI_PLL_CLK/RCIN_RPI_SAMPLE_FREQ)<<12);   // Set pcm div. If we need to configure DMA frequency.
+    hal.scheduler->delay_microseconds(100);
+    clk_reg[RCIN_RPI_PCMCLK_CNTL] = 0x5A000016;                              // Source=PLLD and enable
+    hal.scheduler->delay_microseconds(100);
+    pcm_reg[RCIN_RPI_PCM_TXC_A] = 0<<31 | 1<<30 | 0<<20 | 0<<16;             // 1 channel, 8 bits
+    hal.scheduler->delay_microseconds(100);
+    pcm_reg[RCIN_RPI_PCM_MODE_A] = (10 - 1) << 10;                           //PCM mode 0b0010010000000000
+    hal.scheduler->delay_microseconds(100);
+    pcm_reg[RCIN_RPI_PCM_CS_A] |= 1<<4 | 1<<3;                               // Clear FIFOs
+    hal.scheduler->delay_microseconds(100);
+    pcm_reg[RCIN_RPI_PCM_DREQ_A] = 64<<24 | 64<<8;                           // DMA Req when one slot is free?
+    hal.scheduler->delay_microseconds(100);
+    pcm_reg[RCIN_RPI_PCM_CS_A] |= 1<<9;                                      // Enable DMA
+    hal.scheduler->delay_microseconds(100);
+    pcm_reg[RCIN_RPI_PCM_CS_A] |= 1<<2;                                      // Enable Tx
+    hal.scheduler->delay_microseconds(100);
+}
+
+
+/*Initialise PCM
+  See BCM2711 documentation:
+ */
+void RCInput_RPI::init_PCM_BCM2711()
+{
+    pcm_reg[RCIN_RPI_PCM_CS_A] = 1;                                          // Disable Rx+Tx, Enable PCM block
+    hal.scheduler->delay_microseconds(100);
+    clk_reg[RCIN_RPI_PCMCLK_CNTL] = 0x5A000006;                              // Source=PLLD (700MHz) // https://www.raspberrypi.org/forums/viewtopic.php?t=251902
+    hal.scheduler->delay_microseconds(100);
+    clk_reg[RCIN_RPI_PCMCLK_DIV] = 0x5A000000 | ((RCIN_RPI4_PLL_CLK/RCIN_RPI_SAMPLE_FREQ)<< 12); // 0x5A001900; //0x0x5A001900; //0x5A000000 | ((50000/RCIN_RPI_SAMPLE_FREQ)<< 4);   // Set pcm div. If we need to configure DMA frequency.
```

Comment:
instead of having 2 methods doing basically the exact same thing, why don't we have a single one with:

```c++
if (_version >= 4) {
    pcmclk_div = 0x5A000000 | (RCIN_RPI4_PLL_CLK/RCIN_RPI_SAMPLE_FREQ << 12);
} else {
    pcmclk_div = 0x5A000000 | (RCIN_RPI4_PLL_CLK/RCIN_RPI_SAMPLE_FREQ << 12);
}

clk_reg[RCIN_RPI_PCMCLK_DIV] = pcmclk_div;
```

Reply:
True ... It used two functions for testing and compare...Done.